### PR TITLE
Else check

### PR DESCRIPTION
**Pull request by Oakheart27**
Else Branch

### CODE REVIEW

**Review by evdru — COMMENTED** [review]
Using standard libraries such as stdbool.h, cstring.h, cctype.h.

Maybe put function declarations in a seperate header file.

Have a separate driver file where main is located to bootstrap the program.

Recommend to have a make file to compile and run your program. As well as include a README with instructions on how to compile and run your program.

Include simple unit tests to make sure you handle the base case and edge cases.

**Comment by evdru** [comment]
Maybe put function declarations in a seperate header file.

**Comment by evdru** [comment]
Have a separate driver file where main is located to bootstrap the program.

**Comment by evdru** [comment]
Recommend to have a make file to compile and run your program. As well as include a README with instructions on how to compile and run your program.

**Comment by evdru** [comment]
Include simple unit tests to make sure you handle the base case and edge cases.

File: blah.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    while (numString[currentIndex] != '\0')
+    {
+      mantissa_value_string[currentIndex - offset] = numString[currentIndex];
+      currentIndex++;
+    }
+
+    numerator = atoi(mantissa_value_string);
+    //printf("The mantissa numerator is: %i\n", numerator);
+
+    return numerator;
```
Comment (evdru) [inline blah.cpp]: bool mantissa() returns type int* and not bool?

**Comment by clockian-NAU** [comment]
Just to expand off of what Evan has said.

* "Using standard libraries such as stdbool.h, cstring.h, cctype.h."
Considering this is a very limited embedded system this program is to run on, this is a very reasonable assumption to make. You may need to inquire as to the specific capabilities of the system before using even standard libraries. That being said, this is also a good idea to start development, so you have more tools to work with, and if you need to remove a tool, you know what you need to replace it with. Thanks for that idea.

* "Maybe put function declarations in a separate header file."
For me personally, I use header files as a quick reference, rather than looking at the code itself, so I find them very useful.

* "Have a separate driver file where main is located to bootstrap the program."
This is just standard good practice, but it would also allow you to utilize the OO nature of C++ more easily, a very powerful tool.

* "Recommended to have a make file to compile and run your program. As well as include a README with instructions on how to compile and run your program."
Compilation instructions are nice(personally I had the problem of forgetting the .cpp >.<) but makefiles are great! If you want to get fancy, you can have your project build with just make, and have a separate testing suite build with make test. It can really speed development within your team if you make it super easy for everyone to compile on a whim.

* "Include simple unit tests to make sure you handle the base case and edge cases."
This is a personal thing with me and C, but I find I can't write good, or at least usable, code without tests. Granted, your mileage may vary, and given the limited scope of this project, it may not be necessary, but it is something to consider. It doesn't even have to be anything fancy, it could just be a bunch of functions to print the output of your program with different inputs.

**Comment by clockian-NAU** [comment]
As for the code...

A file header would be nice, as sometimes I work on Vim, a cmd line editor, due to the limited capabilities of my laptop, and it doesn't show the file name I'm working on very easily. It doesn't need to be fancy, but the following would be appreciated.
```c++
/*
 * @file <file-name>
 *
 * @author [maybe, not really necessary for this project I think]
 *
 * @brief <small description of the purpose of the file>
 */
```

**Comment by clockian-NAU** [comment]
With regards to comments, if would be great if you could have them follow a standard format for the code. Below is what I personally prefer, but as long as all of them follow a similar format, anything works honestly.
```C++
// Capitalized first letter with a space from the double slashes

someExampleCode(cool);  // Two spaces away from the code for inline

**Comment by clockian-NAU** [comment]
With regards to the variable naming, good job on having descriptive variables names. It would be great if you could change your variable names to all be either snake case or camel case though, this is some consistency issues with that.

**Comment by clockian-NAU** [comment]
With regards to the program's output, you are missing a requirement where you need to output the denominator in the mantissa function, as it only outputs the numerator at the moment.

**Comment by clockian-NAU** [comment]
Last of the general project and code remarks, I took this opportunity to run the tests I constructed for this assignment and more on your code, below are the results

## Tests that are suppose to succeed

#### Successful tests, tests that passed
```C
"123.456"
"-123.456"
"    -2.3"
"-0002.3"
"    -0002.3"
"2.3      "
```
#### Error'd tests, that are suppose to pass
```C
"2.0"
"2."
"2" - This causes stack smashing
"-2"
"0.3"
"    0.3"
".3"
"2."
"-.3"
```
#### Tests that output wrong numbers
```
"2.3000" -> outputs: c = 2, n = 3000 | n should be 3
"2.3000    " -> see above ^
```

## Tests that are suppose to Fail

#### Successful test, tests that error'd correctly
```
"a.1"
"-a.1"
"-0a.1"
"1.a1"
"1.a0"
"     "
"" -> This exits the program with no output, 
        I don't know whether this is intended behavior or not
```

#### Failed tests, tests that output when they are suppose to fail
```
"-1a.1" -> c = -1, n = 1
"1.1a" -> c = 1, n = 1
"1.1.1" -> c = 1, n = 1
"1.+1" -> c = 1, n = 1
"1.-1" -> c = 1, n = -1
```

File: blah.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        if(isalpha(number))
+        {      
+               printf("test false \n");
+        }
+        
+        // Another solution
+        // Checks if there is another symbol besides a digit or a + - symbol
+        for (int i = 0; i < strlen(number); i = i + 1)
+        { 
+          if(number[i] >= '0' || number[i] <= '9' && number[i] != '+' && number[i] != '-')
```
Comment (clockian-NAU) [inline blah.cpp]: You may want to consider splitting the multiple conditions apart so that it is easier to read at a glance. You may want to reconsider the purpose of this while loop and consider if having a separate loop for '-' and '+' maybe better. You may want to consider if testing for '+' is a part of the requirements of the project

File: blah.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+          }
+        }
+        return 0; 
+    }
+}    
+    
+bool characteristic(char numString[], int& c)
+{
+    //turn char into int
+    c = atoi(numString);
```
Comment (clockian-NAU) [inline blah.cpp]: While I do like the simplicity of this function, and while I am surprised at the mileage atoi() gives to this function, it is a bit to simple for this assignment, failing to cover some edge cases.

File: blah.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        currentIndex++;
+    }
+
+    currentIndex++;
+    offset = currentIndex;
+
+    while (numString[currentIndex] != '\0')
+    {
+      mantissa_value_string[currentIndex - offset] = numString[currentIndex];
+      currentIndex++;
```
Comment (clockian-NAU) [inline blah.cpp]: The indentation in this loop is off by 2 spaces

File: blah.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        mantissa_buffer[currentIndex] = numString[currentIndex];
+        currentIndex++;
+    }
+
+    currentIndex++;
+    offset = currentIndex;
+
+    while (numString[currentIndex] != '\0')
+    {
+      mantissa_value_string[currentIndex - offset] = numString[currentIndex];
```
Comment (clockian-NAU) [inline blah.cpp]: Consider removing the obfuscating math, as the currentIndex minus the offset just counts from 0 to the life of the loop. Might as well us an i++

File: blah.cpp (hunk 1 above)
Comment (clockian-NAU) [reply blah.cpp]: Just to clarify this a bit further, it seems like C++ is similar to C in that bools don't actually exist. It's just an int that holds either 0 or 1. It also seems like the bool can, maybe implicitly, be converted to a number.

Given the above, it seems like in the line
```numerator = atoi(mantissa_value_string);```
This assigns the output to the reference parameter numerator, which acts kind of like a pointer in C, but with more restrictions on it, as C++ is designed(as far as I know) to de-emphasize pointer manipulation that is rampant in C. For this function, this is how the values are returned out of the function, and how the denominator is also suppose to be returned.

Now moving to 
```return numerator;```
What's happening here is that you're trying to return the reference parameter out of a function with the bool return type. This should error, but what seems to be happening is that the return dereferences the numerator and looks at the value it holds, which is an int. It takes the int and tries to fit it in a bool, which is just an int that can only be 0 or 1, thus the number gets reduced down to 1 which is then returned.

I recommend refactoring this code so that the return is a separate variable that solely determines if the function error'd or not, and just passing out the numerator and denominator values through the reference parameters. This also applies to the characteristic function.

File: blah.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    char number[] = "123.456";
+    int c, n, d;
+
+    //if the conversion from C string to integers can take place
+    if(characteristic(number, c) && mantissa(number, n, d))
+    {
+        //do some math with c, n, and d
+        printf("test true \n");
+        printf("The characteristic is: %i\n", c);
+        printf("The mantissa numerator is: %i\n", n);
```
Comment (clockian-NAU) [inline blah.cpp]: I originally thought printf() shouldn't work in C++, since it's C, but it turns out I'm wrong. I would personally advocate for using std::cout just because this is C++, and you aren't outputting enough IO for the performance issue with cout to matter. That being said, here the interesting StackOverflow thread that showed me this. https://stackoverflow.com/questions/2872543/printf-vs-cout-in-c

File: blah.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+bool characteristic(char numString[], int& c);
+bool mantissa(char numString[], int& numerator, int& denominator);
+
+int main()
+{
+    char number[] = "123.456";
+    int c, n, d;
+
+    //if the conversion from C string to integers can take place
+    if(characteristic(number, c) && mantissa(number, n, d))
```
Comment (KennySprings4Life) [inline blah.cpp]: You should consider writing separate functions instead of having all of the functionality occur directly within the main function. The main should act more as a driver which calls other functions, instead of handling all of the processing by itself.
For example: 
Main takes the input from the user, but then sends it to another function that "sanitizes" the data by removing trailing white spaces, verifies that the data is indeed valid, and perhaps strips the "-" / "+" sign from the input for later use. The main would then send this "clean" data to the other functions which will then perform the actions that your current mantissa and characteristic functions do.

**Comment by KennySprings4Life** [comment]
> 
> 
> With regards to comments, if would be great if you could have them follow a standard format for the code. Below is what I personally prefer, but as long as all of them follow a similar format, anything works honestly.
> 
> ```c++
> // Capitalized first letter with a space from the double slashes
> 
> someExampleCode(cool);  // Two spaces away from the code for inline
> ```

I would agree with Jasque here. I noticed also some of the comments for this pull request were capitalized, while others were not. Don't forget to proofread your comments, as well as your code.

**Review by morganlovato — APPROVED** [review]
looks good

**Review by jdm568 — APPROVED** [review]
Looks good!

**Review by morganlovato — APPROVED** [review]
Looks good to me too!
